planet/src/main/java/com/google/planet/data/ItineraryGenerator.java

```diff
@@ -18,46 +18,98 @@
 
         Collections.sort(events, Event.SortByOrder);
-        int [][] travelTimeGraph = getTravelTimeGraph(events);
+        Duration [][] travelTimeGraph = getEventsDistanceMatrix(events);
+        if (errorMessage != null) { 
+            Itinerary itinerary = new Itinerary(Arrays.asList(), errorMessage);
+            return itinerary;
+        }
         List<ItineraryItem> items = scheduleItineraryInOrder(events, START, END, travelTimeGraph);
-        return items;
+        Itinerary itinerary = new Itinerary(items, errorMessage);
+        return itinerary;
     }
 
     // Function that creates a graph with the events as vertices and the travelling
     // time between the events as edges.
-    // For the prototype, all travelling times are set to 15 minutes 
-    public int[][] getTravelTimeGraph(List<Event> events) {
+    // The order of this graph will be the same as the order of the events.
+    // For the MVP, real time traffic is NOT used.
+    // TODO: handle invalid addresses
+    private Duration[][] getEventsDistanceMatrix(List<Event> events) {
         int numOfNodes = events.size();
-        int[][] travelTimeGraph = new int[numOfNodes][numOfNodes];
-        for (int i = 0; i < travelTimeGraph.length; i++) {
-            for (int j = 0; j < travelTimeGraph[i].length; j++) {
-                if (i == j) {
-                    travelTimeGraph[i][j] = 0;
-                }else{
-                    travelTimeGraph[i][j] = 15;
+        Duration[][] travelTimeGraph = new Duration[numOfNodes][numOfNodes];
+
+        String GoogleApiKey = "AIzaSyDK36gDoYgOj4AlbCqh1IuaUuTlcpKF0ns";
+        String[] origins = getListOfAddresses(events);
+        String[] destinations = getListOfAddresses(events);
+        GeoApiContext context = new GeoApiContext.Builder(new GaeRequestHandler.Builder())
+                            .apiKey(GoogleApiKey)
+                            .build();
+       try {
+            DistanceMatrix distanceMatrix = DistanceMatrixApi.getDistanceMatrix(context,
+                                                         origins,
+                                                         destinations).await();
+
+            for (int i = 0; i < distanceMatrix.rows.length; i++) {
+                DistanceMatrixRow currentRow = distanceMatrix.rows[i];
+                for (int j = 0; j < currentRow.elements.length; j++) {
+                    DistanceMatrixElement currentElement = currentRow.elements[j];
+                    if (currentElement.duration != null) {
+                        travelTimeGraph[i][j] = currentElement.duration;
+                    } else {
+                        errorMessage = "Oops, one of your addresses is invalid, please use a valid address.";
+                    }
                 }
             }
+            return travelTimeGraph;
+        } catch (Exception e) {
+            errorMessage = "Oops, one of your addresses is invalid, please use a valid address.";
```

Comment:
same for these errors. They should be the appropriate 4XX errors

Reply:
fixed!